### PR TITLE
Fix541

### PR DESCRIPTION
**Pull request by b-scholz**
- Convert RamRelationReference to a proper proxy pattern
- Change references in cloned program to new relations (cf. clone() of RamProgram)

### CODE REVIEW

File: src/RamProgram.h [hunk 1]
```diff
@@ -118,13 +118,27 @@ class RamProgram : public RamNode {
 
     /** Create clone */
     RamProgram* clone() const override {
+        typedef std::map<const RamRelation*, const RamRelation*> RefMapType;
```
Comment (mmcgr) [inline src/RamProgram.h]: using RefMapType =  std::map<const RamRelation*, const RamRelation*>;

Comment (mmcgr) [inline src/RamProgram.h]: Why the 'type' in RefMapType?

File: src/RamProgram.h [hunk 2]
```diff
@@ -118,13 +118,27 @@ class RamProgram : public RamNode {
...
+            refMap[cur.second.get()] = newRel;
+            res->addRelation(std::unique_ptr<RamRelation>(newRel));
         }
         for (auto& cur : subroutines) {
             res->addSubroutine(cur.first, std::unique_ptr<RamStatement>(cur.second->clone()));
         }
+        res->apply(makeLambdaRamMapper([&](std::unique_ptr<RamNode> node) -> std::unique_ptr<RamNode> {
+            // check whether it is a unnamed variable
+            auto* relRef = dynamic_cast<RamRelationReference*>(node.get());
+            if (!relRef) {
```
Comment (mmcgr) [inline src/RamProgram.h]:
```suggestion
            if (relRef != nullptr) {
```

File: src/RamRelation.h [hunk 3]
```diff
@@ -83,6 +84,7 @@ class RamRelation : public RamNode {
         return "c" + std::to_string(i);
     }
 
+    /** Get Argument Type */
```
Comment (mmcgr) [inline src/RamRelation.h]: Argument type, or argument type qualifier?

File: src/RamProgram.h (hunk 1 above)
Comment (b-scholz) [reply src/RamProgram.h]: I grew up with C...

Comment (b-scholz) [reply src/RamProgram.h]: indeed, not needed anymore; before I had a RamMapper class and needed to pass arguments having a replicated use of the type. However, with makeLambdaRamMapper the current environment can be used.

File: src/RamProgram.h (hunk 2 above)
Comment (b-scholz) [reply src/RamProgram.h]: Done.

File: src/RamRelation.h (hunk 3 above)
Comment (b-scholz) [reply src/RamRelation.h]: Done.

**Pull request merged**